# **RNN**
A recurrent neural network (RNN) is a class of artificial neural network where connections between units form a directed cycle. This creates an internal state of the network which allows it to exhibit dynamic temporal behavior.

IMDB sentiment classification task

This is a dataset for binary sentiment classification containing substantially more data than previous benchmark datasets. IMDB provided a set of 25,000 highly polar movie reviews for training, and 25,000 for testing. There is additional unlabeled data for use as well. Raw text and already processed bag of words formats are provided.

You can download the dataset from http://ai.stanford.edu/~amaas/data/sentiment/  or you can directly use 
" from keras.datasets import imdb " to import the dataset.

Few points to be noted:
Modules like SimpleRNN, LSTM, Activation layers, Dense layers, Dropout can be directly used from keras
For preprocessing, you can use required 

In [1]:
#load the imdb dataset 
from keras.datasets import imdb

vocabulary_size = 5000
(X_train, y_train), (X_test, y_test) = imdb.load_data(num_words = vocabulary_size)
print('Loaded dataset with {} training samples, {} test samples'.format(len(X_train), len(X_test)))

17473536/17464789 [==============================] - 0s 0us/step
Loaded dataset with 25000 training samples, 25000 test samples


In [2]:
#the review is stored as a sequence of integers. 
# These are word IDs that have been pre-assigned to individual words, and the label is an integer

print('---review---')
print(X_train[6])
print('---label---')
print(y_train[6])

# to get the actual review
word2id = imdb.get_word_index()
id2word = {i: word for word, i in word2id.items()}
print('---review with words---')
print([id2word.get(i, ' ') for i in X_train[6]])
print('---label---')
print(y_train[6])

---review---
[1, 2, 365, 1234, 5, 1156, 354, 11, 14, 2, 2, 7, 1016, 2, 2, 356, 44, 4, 1349, 500, 746, 5, 200, 4, 4132, 11, 2, 2, 1117, 1831, 2, 5, 4831, 26, 6, 2, 4183, 17, 369, 37, 215, 1345, 143, 2, 5, 1838, 8, 1974, 15, 36, 119, 257, 85, 52, 486, 9, 6, 2, 2, 63, 271, 6, 196, 96, 949, 4121, 4, 2, 7, 4, 2212, 2436, 819, 63, 47, 77, 2, 180, 6, 227, 11, 94, 2494, 2, 13, 423, 4, 168, 7, 4, 22, 5, 89, 665, 71, 270, 56, 5, 13, 197, 12, 161, 2, 99, 76, 23, 2, 7, 419, 665, 40, 91, 85, 108, 7, 4, 2084, 5, 4773, 81, 55, 52, 1901]
---label---
1
1654784/1641221 [==============================] - 0s 0us/step
---review with words---
['the', 'and', 'full', 'involving', 'to', 'impressive', 'boring', 'this', 'as', 'and', 'and', 'br', 'villain', 'and', 'and', 'need', 'has', 'of', 'costumes', 'b', 'message', 'to', 'may', 'of', 'props', 'this', 'and', 'and', 'concept', 'issue', 'and', 'to', "god's", 'he', 'is', 'and', 'unfolds', 'movie', 'women', 'like', "isn't", 'surely', "i'm", 'and', 'to', 'toward', 

In [3]:
#pad sequences (write your code here)
from keras.preprocessing import sequence
max_words = 500
X_train = sequence.pad_sequences(X_train, maxlen=max_words)
X_test = sequence.pad_sequences(X_test, maxlen=max_words)

In [4]:
#design a RNN model (write your code)

from keras import Sequential
from keras.layers import Embedding, LSTM, Dense, Dropout, SimpleRNN

embedding_size=32
model_rnn = Sequential()
model_rnn.add(Embedding(vocabulary_size, embedding_size, input_length=max_words))
model_rnn.add(SimpleRNN(100))
model_rnn.add(Dense(1, activation='sigmoid'))
print(model_rnn.summary())

Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 embedding (Embedding)       (None, 500, 32)           160000    
                                                                 
 simple_rnn (SimpleRNN)      (None, 100)               13300     
                                                                 
 dense (Dense)               (None, 1)                 101       
                                                                 
Total params: 173,401
Trainable params: 173,401
Non-trainable params: 0
_________________________________________________________________
None


In [5]:
#train and evaluate your model
#choose your loss function and optimizer and mention the reason to choose that particular loss function and optimizer
# use accuracy as the evaluation metric

model_rnn.compile(loss='binary_crossentropy', 
             optimizer='adam', 
             metrics=['accuracy'])

Binary cross entropy is chosen due to the nature of the problem i.e. binary classification. Adam optimiser has been proven to be among the best optimisers.

In [7]:
batch_size = 64
num_epochs = 3

X_valid, y_valid = X_train[:batch_size], y_train[:batch_size]
X_train2, y_train2 = X_train[batch_size:], y_train[batch_size:]
model_rnn.fit(X_train2, y_train2, validation_data=(X_valid, y_valid), batch_size=batch_size, epochs=num_epochs)

Epoch 1/3
390/390 [==============================] - 245s 619ms/step - loss: 0.6394 - accuracy: 0.6115 - val_loss: 0.5671 - val_accuracy: 0.7031
Epoch 2/3
390/390 [==============================] - 242s 619ms/step - loss: 0.5718 - accuracy: 0.7060 - val_loss: 0.5284 - val_accuracy: 0.7344
Epoch 3/3
390/390 [==============================] - 241s 618ms/step - loss: 0.5329 - accuracy: 0.7273 - val_loss: 0.6801 - val_accuracy: 0.6094


In [8]:
#evaluate the model using model.evaluate()
scores = model_rnn.evaluate(X_test, y_test, verbose=0)
print('Test accuracy:', scores[1])

Test accuracy: 0.6280800104141235


# **LSTM**

Instead of using a RNN, now try using a LSTM model and compare both of them. Which of those performed better and why ?


In [10]:
from keras import Sequential
from keras.layers import Embedding, LSTM, Dense, Dropout, SimpleRNN

embedding_size=32
model_lstm = Sequential()
model_lstm.add(Embedding(vocabulary_size, embedding_size, input_length=max_words))
model_lstm.add(LSTM(100))
model_lstm.add(Dense(1, activation='sigmoid'))
print(model_lstm.summary())

model_lstm.compile(loss='binary_crossentropy', 
             optimizer='adam', 
             metrics=['accuracy'])

batch_size = 64
num_epochs = 3

X_valid, y_valid = X_train[:batch_size], y_train[:batch_size]
X_train2, y_train2 = X_train[batch_size:], y_train[batch_size:]
model_lstm.fit(X_train2, y_train2, validation_data=(X_valid, y_valid), batch_size=batch_size, epochs=num_epochs)

scores = model_lstm.evaluate(X_test, y_test, verbose=0)
print('Test accuracy:', scores[1])

Model: "sequential_2"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 embedding_2 (Embedding)     (None, 500, 32)           160000    
                                                                 
 lstm_1 (LSTM)               (None, 100)               53200     
                                                                 
 dense_2 (Dense)             (None, 1)                 101       
                                                                 
Total params: 213,301
Trainable params: 213,301
Non-trainable params: 0
_________________________________________________________________
None
Epoch 1/3
390/390 [==============================] - 35s 78ms/step - loss: 0.4554 - accuracy: 0.7780 - val_loss: 0.2900 - val_accuracy: 0.8750
Epoch 2/3
390/390 [==============================] - 30s 77ms/step - loss: 0.2883 - accuracy: 0.8851 - val_loss: 0.2262 - val_accuracy: 0.9062
Epoch 3/3
390/390 [==

LSTM performs considerably better than RNN. 62.8% test accuracy in RNN vs 86.69% test accuracy in LSTM. This was expected as LSTM solves the problem of vanishing and exploding gradients(or long range dependencies) and consists of 3 gates whereas vanilla RNN has just 1 gate.

Perform Error analysis and explain using few examples.

In [11]:
# for user prediction
def user_input_processing(model, review):
    vec = []
    for word in review.split(" "):
        if word[-1] == ".":
            word = word[:-1]
        vec.append(word2id[str.lower(word)])
    vec_padded = sequence.pad_sequences([vec], 500)
    print(review, model.predict(vec_padded))

In [12]:
user_input_processing(model_rnn, "One of the finest films made in recent years.")

One of the finest films made in recent years. [[0.56504375]]


In [17]:
user_input_processing(model_rnn, "Predictable and bad. The acting was terrible and the story was common.")

Predictable and bad. The acting was terrible and the story was common. [[0.3047255]]


In [14]:
user_input_processing(model_lstm, "One of the finest films made in recent years.")

One of the finest films made in recent years. [[0.8364623]]


In [16]:
user_input_processing(model_lstm, "Predictable and bad. The acting was terrible and the story was common.")

Predictable and bad. The acting was terrible and the story was common. [[0.5422326]]


On the positive review "One of the finest films made in recent years.", LSTM gives higher score than RNN but on the negative review RNN has a lower score than LSTM. This can be due to the model being trained on very small data.